docs/csharp/language-reference/operators/null-coalescing-operator.md

```diff
@@ -25,7 +28,25 @@ is evaluated as
 a ?? (b ?? c)
 ```
 
-The `??` operator can be useful in the following scenarios:
+In C# 7.3 and earlier, the type of the left-hand operand must be either a reference type or a [nullable value type](../../programming-guide/nullable-types/index.md). Beginning with C# 8.0, that requirement is replaced with the following: the type of the left-hand operand cannot be a non-nullable value type. In particular, you can use the null-coalescing operator with unconstrained type parameters:
+
+[!code-csharp[unconstrained type parameter](~/samples/csharp/language-reference/operators/NullCoalescingOperator.cs#UnconstrainedType)]
+
+## Null-coalescing assignment
+
+Available in C# 8.0 and later, the null-coalescing assignment operator `??=` assigns the value of its right-hand operand to its left-hand operand only if the left-hand operand is null. The `??=` operator doesn't evaluate its right-hand operand if the left-hand operand evaluates to non-null.
+
+[!code-csharp[null-coalescing assignment](~/samples/csharp/language-reference/operators/NullCoalescingOperator.cs#Assignment)]
+
+Unlike with [compound assignment operators](assignment-operator.md#compound-assignment), the type of the result of the `??=` operator may differ from the type of its left-hand operand. For expression `a ??= b`, if `a` is of a nullable value type `T?` and `b` is implicitly convertible to the corresponding underlying type `T`, the type of the result of `a ??= b` is `T`. The null-coalescing operator `??` also behaves that way.
```

Comment:
This wording troubles me. But, I haven't come up with a better way to explain it. I'll discussmy concerns, so we can discuss together:

The result type *is* the same as the left hand side argument. However, it is guaranteed to be non-null (other than the pathological case of `a ??= null;`).  So, introducing the concept of *type* might be confusing. For example:

```csharp
int? a = null;
int b = 5;
a ??= b; // a is still an int?, but it's HasValue property is true.
```

By saying "the result of the `??=` operator may differ..." that could lead readers to assume `a` is an `int`, not an `int?`.

Reply:
@BillWagner 

> The result type _is_ the same as the left hand side argument.

What do you mean by the "result type"? If the type of the result returned by the `??=` operator, then it isn't always the same as the left-hand argument. The following code shows that:

```csharp
var xs = new List<int>();
int? a = 5;
xs.Add(a ??= 10);  \\ that compiles because ??= returns an `int` value
xs.Add(a);  \\ that doesn't compile because a is `int?`
```

That's why I mentioned, it's "unlike with compound assignment operators".

> By saying "the result of the `??=` operator may differ..." that could lead readers to assume `a` is an `int`, not an `int?`.

I agree with that possible confusion. Is that the "dangerous" confusion though? That would mean that a variable *has changed* its type as a result of the operation.

Though, what about the following wording:

> If the left-hand operand is of the nullable value type `T?`, expression `a ??= b` can return the value of the corresponding underlying type `T`. That happens when the type of `b` is implicitly convertible to type `T`.

Reply:
@BillWagner another option is to remove that paragraph at all. Technically interested readers would follow the [feature proposal link](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/null-coalescing-assignment) where all the details are mentioned.

I think that would simplify the article.